File: projects/sdk-kotlin/src/main/kotlin/orca/sdk/Transport.kt
// TCP + mTLS transport: plugin (client) side. Wire-compatible with
// projects/sdk/src/transport.rs, projects/sdk-go/transport, and
// projects/sdk-ts/src/transport.ts.
package orca.sdk

import java.io.DataInputStream
import java.net.InetSocketAddress
import java.nio.file.Path
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong
import javax.net.ssl.SSLSocket
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.consumeAsFlow
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.launch
import kotlinx.coroutines.runInterruptible
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long
import kotlinx.serialization.json.put

const val SDK_VERSION: String = "0.1.0"
const val CONTEXT_EVENT_METHOD: String = "orca/context.event"

enum class Flavor(val wire: String) {
    FULL("full"),
    HEADLESS("headless"),
    LOCAL("local");
}

enum class Sensitivity(val wire: String) {
    GENERAL("general"),
    SENSITIVE("sensitive");
}

/**
 * Builder for [Transport.helloFull]. Mirrors HelloOptions in Rust/Go/TS —
 * not @Serializable; the wire shape is the inline JSON object built in
 * helloFull(). This data class is the ergonomic input.
 */
data class HelloOptions(
    val pluginId: String,
    val pluginVersion: String = "",
    val flavor: Flavor = Flavor.HEADLESS,
    val coreMinRequired: String = "0.1.0",
    val methodsRequired: List<String> = emptyList(),
    val methodsOptional: List<String> = emptyList(),
    val pluginsRequired: List<String> = emptyList(),
    val pluginsOptional: List<String> = emptyList(),
)

@Serializable
data class HelloResult(
    val server_version: String,
    val ok: Boolean,
    val status: String,
    val methods: List<String>,
    val reason: String? = null,
)

@Serializable
data class TypesDeclareResult(val accepted: List<String>)

@Serializable
data class TypedValue(
    val type: String,
    val schema_version: String,
    val sensitivity: String,
    val payload: JsonElement,
)

@Serializable
data class ContextEvent(
    val subscription_id: String,
    val context_id: String,
    val value: TypedValue,
)

private val json = Json {
    encodeDefaults = true
    ignoreUnknownKeys = true
    explicitNulls = false
}

class Transport private constructor(
    private val socket: SSLSocket,
    private val parentJob: Job,
) {
    private val scope = CoroutineScope(SupervisorJob(parentJob) + Dispatchers.IO)
    private val output = socket.outputStream
    private val input = DataInputStream(socket.inputStream)
    private val nextId = AtomicLong(1)
    private val pending = ConcurrentHashMap<Long, CompletableDeferred<Response>>()
    private val notifFlow = MutableSharedFlow<Notification>(extraBufferCapacity = 256)
    private val writeMutex = Object()

    /** Tools registered for the host to invoke, keyed by bare name. */
    private val tools = ConcurrentHashMap<String, RegisteredTool>()

    init {
        scope.launch { readLoop() }
    }

    companion object {
        /**
         * Connect to host:port over mTLS using `bundle`. Suspends until the
         * TLS handshake completes.
         */
        suspend fun connect(addr: String, bundle: NodeBundle): Transport {
            val (host, portStr) = addr.split(":", limit = 2).let {
                require(it.size == 2) { "invalid addr: $addr" }
                it[0] to it[1]
            }
            val sslContext = Pki.clientSslContext(bundle)
            val factory = sslContext.socketFactory
            return runInterruptible(Dispatchers.IO) {
                val raw = java.net.Socket()
                raw.connect(InetSocketAddress(host, portStr.toInt()), 10_000)
                val ssl = factory.createSocket(raw, "core.orca.local", portStr.toInt(), true) as SSLSocket
                ssl.useClientMode = true
                ssl.startHandshake()
                Transport(ssl, Job())
            }
        }
    }

    fun close() {
        try {
            socket.close()
        } catch (_: Exception) {}
        scope.cancel()
        for ((_, def) in pending) {
            def.complete(
                Response(
                    id = null,
                    error = ErrorObject(CODE_INTERNAL_ERROR, "transport closed"),
                )
            )
        }
        pending.clear()
    }

    /** Flow of every server-pushed notification. */
    fun notifications(): Flow<Notification> = notifFlow.asSharedFlow()

    private suspend fun readLoop() {
        try {
            while (true) {
                val frame = runInterruptible { Framing.read(input) } ?: break
                val raw = json.parseToJsonElement(String(frame, Charsets.UTF_8)).jsonObject
                val hasMethod = raw.containsKey("method")
                val hasId = raw.containsKey("id")
                val hasResultOrError = raw.containsKey("result") || raw.containsKey("error")
                when {
                    hasId && hasResultOrError && !hasMethod -> {
                        val resp = json.decodeFromJsonElement(Response.serializer(), raw)
                        val id = resp.id
                        if (id != null) pending.remove(id)?.complete(resp)
                    }
                    hasMethod && !hasId -> {
                        val notif = json.decodeFromJsonElement(Notification.serializer(), raw)
                        notifFlow.tryEmit(notif)
                    }
                    hasMethod && hasId -> {
                        // Server→plugin request. Dispatch off the read loop
                        // so a slow handler doesn't stall incoming frames.
                        val req = json.decodeFromJsonElement(Request.serializer(), raw)
                        scope.launch { dispatchIncoming(req) }
                    }
                }
            }
        } catch (_: Exception) {
            // Stream closed or malformed — fall through to close.
        } finally {
            close()
        }
    }

    private fun writeBytes(body: ByteArray) {
        synchronized(writeMutex) { Framing.write(output, body) }
    }

    /** Send a request and suspend until the matching response arrives. */
    suspend fun call(method: String, params: JsonElement?): Response {
        val id = nextId.getAndIncrement()
        val def = CompletableDeferred<Response>()
        pending[id] = def
        val req = buildJsonObject {
            put("jsonrpc", JSONRPC_VERSION)
            put("id", id)
            put("method", method)
            if (params != null) put("params", params)
        }
        try {
            writeBytes(json.encodeToString(JsonObject.serializer(), req).toByteArray())
        } catch (e: Exception) {
            pending.remove(id)
            throw e
        }
        return def.await()
    }

    suspend fun hello(
        pluginId: String,
        flavor: Flavor = Flavor.HEADLESS,
        methodsRequired: List<String> = emptyList(),
        methodsOptional: List<String> = emptyList(),
    ): HelloResult = helloFull(
        HelloOptions(
            pluginId = pluginId,
            flavor = flavor,
            methodsRequired = methodsRequired,
            methodsOptional = methodsOptional,
        )
    )

    /**
     * Full orca/hello with peer-plugin dependencies and own version.
     * Use this when porting an `orca-plugin.toml` straight through.
     */
    suspend fun helloFull(opts: HelloOptions): HelloResult {
        val params = buildJsonObject {
            put("sdk_version", SDK_VERSION)
            put("plugin_id", opts.pluginId)
            if (opts.pluginVersion.isNotEmpty()) put("plugin_version", opts.pluginVersion)
            put("flavor", opts.flavor.wire)
            put("core_min_required", opts.coreMinRequired)
            if (opts.methodsRequired.isNotEmpty())
                put("methods_required", json.encodeToJsonElement(opts.methodsRequired))
            if (opts.methodsOptional.isNotEmpty())
                put("methods_optional", json.encodeToJsonElement(opts.methodsOptional))
            if (opts.pluginsRequired.isNotEmpty())
                put("plugins_required", json.encodeToJsonElement(opts.pluginsRequired))
            if (opts.pluginsOptional.isNotEmpty())
                put("plugins_optional", json.encodeToJsonElement(opts.pluginsOptional))
        }
        val resp = call("orca/hello", params)
        if (resp.isError) error("orca/hello rejected: ${resp.error?.message}")
        val result = json.decodeFromJsonElement(HelloResult.serializer(), resp.result!!)
        if (!result.ok) error("orca/hello: server returned ok=false (status=${result.status}; ${result.reason ?: "no reason"})")
        return result
    }

    /**
     * Forward a tool call to a peer plugin via the host. `fqName` is
     * `<peer>.<tool>`. The host resolves the owning plugin, dispatches
     * tools.call, and returns the peer's opaque result. `timeoutMs` is
     * forwarded to the host (rounded to whole seconds).
     */
    suspend fun invokeTool(
        fqName: String,
        arguments: JsonElement,
        timeoutMs: Long = 30_000L,
    ): JsonElement {
        val params = buildJsonObject {
            put("name", fqName)
            put("arguments", arguments)
            put("timeout_secs", maxOf(1L, timeoutMs / 1000L))
        }
        val resp = call(ToolsProtocol.INVOKE_METHOD, params)
        if (resp.isError) error("orca/tools.invoke $fqName failed: ${resp.error?.message}")
        return json.decodeFromJsonElement(ToolInvokeResult.serializer(), resp.result!!).result
    }

    /** Ask the host which peer plugins are currently connected. */
    suspend fun listPeers(): List<PeerInfo> {
        val resp = call(ToolsProtocol.PLUGINS_LIST_METHOD, buildJsonObject {})
        if (resp.isError) error("orca/plugins.list: ${resp.error?.message}")
        return json.decodeFromJsonElement(PluginsListResult.serializer(), resp.result!!).peers
    }

    suspend fun declareTypes(types: List<TypeDeclaration>): TypesDeclareResult {
        val params = buildJsonObject {
            put("types", json.encodeToJsonElement(types))
        }
        val resp = call("orca/types.declare", params)
        if (resp.isError) error("orca/types.declare rejected: ${resp.error?.message}")
        return json.decodeFromJsonElement(TypesDeclareResult.serializer(), resp.result!!)
    }

    suspend fun publishContext(contextId: String, value: TypedValue) {
        val params = buildJsonObject {
            put("context_id", contextId)
            put("value", json.encodeToJsonElement(value))
        }
        val resp = call("orca/context.publish", params)
        if (resp.isError) error("orca/context.publish rejected: ${resp.error?.message}")
    }

    suspend fun subscribeContext(
        contextId: String,
        typeFilter: List<String> = emptyList(),
    ): Pair<String, Flow<ContextEvent>> {
        val params = buildJsonObject {
            put("context_id", contextId)
            put("type_filter", json.encodeToJsonElement(typeFilter))
        }
        val resp = call("orca/context.subscribe", params)
        if (resp.isError) error("orca/context.subscribe rejected: ${resp.error?.message}")
        val subId = resp.result!!.jsonObject["subscription_id"]!!.jsonPrimitive.toString().trim('"')

        val events = Channel<ContextEvent>(Channel.BUFFERED)
        scope.launch {
            notifFlow.filter { it.method == CONTEXT_EVENT_METHOD }.collect { n ->
                val ev = json.decodeFromJsonElement(ContextEvent.serializer(), n.params!!)
                if (ev.subscription_id == subId) events.send(ev)
            }
        }
        return subId to events.consumeAsFlow()
    }

    suspend fun unsubscribeContext(subscriptionId: String) {
        val params = buildJsonObject { put("subscription_id", subscriptionId) }
        val resp = call("orca/context.unsubscribe", params)
        if (resp.isError) error("orca/context.unsubscribe rejected: ${resp.error?.message}")
    }

    // ── Tools surface ─────────────────────────────────────────────────────

    /**
     * Register a tool the host can invoke via orca/tools.call. Bare name
     * (no `<plugin_id>.` prefix — the host applies the namespace).
     * Re-registering the same name replaces the previous handler. Call
     * this for each tool, then call [declareTools] once.
     */
    fun registerTool(
        name: String,
        description: String,
        inputSchema: JsonElement,
        sensitivity: Sensitivity,
        handler: ToolHandler,
    ) {
        val decl = ToolDeclaration(
            name = name,
            description = description,
            input_schema = inputSchema,
            sensitivity = sensitivity.wire,
        )
        tools[name] = RegisteredTool(decl, handler)
    }

    /**
     * Send the registered tool set via orca/tools.declare. Returns the
     * namespaced ids the host accepted. Idempotent — calling again
     * replaces the host-side set.
     */
    suspend fun declareTools(): ToolsDeclareResult {
        val decls = tools.values.map { it.declaration }
        val params = buildJsonObject {
            put("tools", json.encodeToJsonElement(decls))
        }
        val resp = call(ToolsProtocol.DECLARE_METHOD, params)
        if (resp.isError) error("${ToolsProtocol.DECLARE_METHOD} rejected: ${resp.error?.message}")
        return json.decodeFromJsonElement(ToolsDeclareResult.serializer(), resp.result!!)
    }

    private suspend fun dispatchIncoming(req: Request) {
        val resp = buildResponseFor(req)
        try {
            writeBytes(json.encodeToString(Response.serializer(), resp).toByteArray())
        } catch (_: Exception) {
            // Connection probably closed; nothing to do.
        }
    }

    private suspend fun buildResponseFor(req: Request): Response {
        if (req.method != ToolsProtocol.CALL_METHOD) {
            return Response(
                id = req.id,
                error = ErrorObject(CODE_METHOD_NOT_FOUND, "method not found: ${req.method}"),
            )
        }
        val params = req.params
            ?: return Response(
                id = req.id,
                error = ErrorObject(CODE_INVALID_PARAMS, "invalid params: missing params"),
            )
        val callParams = try {
            json.decodeFromJsonElement(ToolCallParams.serializer(), params)
        } catch (e: Exception) {
            return Response(
                id = req.id,
                error = ErrorObject(CODE_INVALID_PARAMS, "invalid params: ${e.message}"),
            )
        }
        val reg = tools[callParams.name]
            ?: return Response(
                id = req.id,
                error = ErrorObject(
                    ToolErrorCodes.UNKNOWN_TOOL,
                    "unknown tool: ${callParams.name}",
                ),
            )
        return try {
            val out = reg.handler.call(callParams.arguments)
            val result = json.encodeToJsonElement(ToolCallResult(result = out))
            Response(id = req.id, result = result)
        } catch (e: ToolHandlerError) {
            Response(
                id = req.id,
                error = ErrorObject(ToolErrorCodes.HANDLER_ERROR, e.message ?: "", e.data),
            )
        } catch (e: Exception) {
            Response(id = req.id, error = ErrorObject(CODE_INTERNAL_ERROR, e.message ?: e.toString()))
        }
    }
}

@Serializable
data class TypeDeclaration(
    val type_name: String,
    val schema_version: String,
    val schema: JsonElement,
    val sensitivity: String,
)

private inline fun <reified T> Json.encodeToJsonElement(value: T): JsonElement =
    parseToJsonElement(encodeToString(value))

/** Convenience: load a bundle from `pkiDir` and connect. */
suspend fun connectAsPlugin(addr: String, pkiDir: Path, pluginId: String): Transport {
    val bundle = Pki.loadPlugin(pkiDir, pluginId)
    return Transport.connect(addr, bundle)
}

File: projects/sdk-ts/src/transport.ts
/**
 * TCP + mTLS transport: plugin (client) side. Wire-compatible with
 * projects/sdk/src/transport.rs and projects/sdk-go/transport.
 *
 * One Transport wraps a mutually-authenticated TLS stream. A reader loop
 * demuxes incoming frames:
 *   - Responses are routed to the matching call() promise via a per-id table.
 *   - Notifications fan out to subscribers registered via notifications().
 */
import { connect as tlsConnect, type TLSSocket } from 'node:tls';
import { FrameReader, writeFrame } from './framing.js';
import {
  classifyMessage,
  type ErrorObject,
  JSONRPC_VERSION,
  type Notification,
  type Request,
  type Response,
} from './jsonrpc.js';
import { clientTlsOptions, type NodeBundle } from './pki.js';
import {
  type RegisteredTool,
  TOOLS_CALL_METHOD,
  TOOLS_DECLARE_METHOD,
  type ToolCallParams,
  type ToolCallResult,
  type ToolDeclaration,
  ToolHandlerError,
  type ToolHandler,
  type ToolsDeclareParams,
  type ToolsDeclareResult,
  toolErrorCodes,
} from './tools.js';

export const SDK_VERSION = '0.1.0';

export type Flavor = 'full' | 'headless' | 'local';
export type Sensitivity = 'general' | 'sensitive';

export interface HelloParams {
  sdk_version: string;
  plugin_id: string;
  flavor: Flavor;
  core_min_required: string;
  methods_required: string[];
  methods_optional: string[];
}

export interface HelloResult {
  server_version: string;
  ok: boolean;
  status: 'full' | 'degraded' | 'rejected' | string;
  methods: string[];
  reason?: string;
}

export interface TypeDeclaration {
  type_name: string;
  schema_version: string;
  schema: unknown;
  sensitivity: Sensitivity;
}

export interface TypesDeclareResult {
  accepted: string[];
}

export interface TypedValue {
  type: string;
  schema_version: string;
  sensitivity: Sensitivity;
  payload: unknown;
}

export interface ContextEvent {
  subscription_id: string;
  context_id: string;
  value: TypedValue;
}

export const CONTEXT_EVENT_METHOD = 'orca/context.event';

export interface ConnectOptions {
  /** Host:port (e.g. "127.0.0.1:5051"). */
  addr: string;
  bundle: NodeBundle;
  /** Optional handshake timeout in ms (default 10s). */
  timeoutMs?: number;
}

export class Transport {
  private writer: TLSSocket;
  private reader: FrameReader;
  private nextID = 1;
  private pending = new Map<number, (resp: Response) => void>();
  private notifSubs = new Set<(n: Notification) => void>();
  /** Tools the plugin has registered for the host to invoke, by bare name. */
  private tools = new Map<string, RegisteredTool>();
  private closed = false;

  private constructor(socket: TLSSocket) {
    this.writer = socket;
    this.reader = new FrameReader(socket);
    socket.on('close', () => this.handleClose(new Error('transport closed')));
    socket.on('error', err => this.handleClose(err));
    void this.readLoop();
  }

  static connect(opts: ConnectOptions): Promise<Transport> {
    const [host, portStr] = opts.addr.split(':');
    if (!host || !portStr) {
      return Promise.reject(new Error(`invalid addr: ${opts.addr}`));
    }
    const port = Number(portStr);
    return new Promise((resolve, reject) => {
      const tlsOpts = clientTlsOptions(opts.bundle);
      const socket = tlsConnect(port, host, tlsOpts);
      const timeout = setTimeout(() => {
        socket.destroy();
        reject(new Error(`tls handshake timeout for ${opts.addr}`));
      }, opts.timeoutMs ?? 10_000);
      socket.once('secureConnect', () => {
        clearTimeout(timeout);
        resolve(new Transport(socket));
      });
      socket.once('error', err => {
        clearTimeout(timeout);
        reject(err);
      });
    });
  }

  private handleClose(err: Error): void {
    if (this.closed) return;
    this.closed = true;
    for (const resolver of this.pending.values()) {
      resolver({
        jsonrpc: JSONRPC_VERSION,
        id: null,
        error: { code: -32603, message: err.message },
      });
    }
    this.pending.clear();
    this.notifSubs.clear();
  }

  private async readLoop(): Promise<void> {
    try {
      while (true) {
        const frame = await this.reader.read();
        if (frame === null) {
          this.handleClose(new Error('connection closed'));
          return;
        }
        let parsed: unknown;
        try {
          parsed = JSON.parse(frame.toString('utf8'));
        } catch {
          continue;
        }
        let msg;
        try {
          msg = classifyMessage(parsed);
        } catch {
          continue;
        }
        if (msg.kind === 'response') {
          const id = msg.value.id;
          if (typeof id === 'number') {
            const resolver = this.pending.get(id);
            if (resolver) {
              this.pending.delete(id);
              resolver(msg.value);
            }
          }
        } else if (msg.kind === 'notification') {
          for (const sub of [...this.notifSubs]) {
            try {
              sub(msg.value);
            } catch {
              /* swallow subscriber errors */
            }
          }
        } else if (msg.kind === 'request') {
          // Server→plugin request. Dispatch off the read loop so a slow
          // handler doesn't stall incoming frames.
          void this.dispatchIncoming(msg.value);
        }
      }
    } catch (err) {
      this.handleClose(err as Error);
    }
  }

  /** Close the underlying socket. Pending calls reject. */
  close(): void {
    this.writer.end();
  }

  /**
   * Handle a server→plugin request. Currently only `orca/tools.call` is
   * supported; everything else returns method-not-found.
   */
  private async dispatchIncoming(req: Request): Promise<void> {
    const resp = await this.buildResponseFor(req);
    try {
      const body = Buffer.from(JSON.stringify(resp), 'utf8');
      await writeFrame(this.writer, body);
    } catch {
      /* swallow — connection likely closed */
    }
  }

  private async buildResponseFor(req: Request): Promise<Response> {
    if (req.method !== TOOLS_CALL_METHOD) {
      return {
        jsonrpc: JSONRPC_VERSION,
        id: req.id,
        error: { code: -32601, message: `method not found: ${req.method}` },
      };
    }
    const params = req.params as ToolCallParams | undefined;
    if (!params || typeof params.name !== 'string') {
      return {
        jsonrpc: JSONRPC_VERSION,
        id: req.id,
        error: { code: -32602, message: 'invalid params: missing name' },
      };
    }
    const reg = this.tools.get(params.name);
    if (!reg) {
      return {
        jsonrpc: JSONRPC_VERSION,
        id: req.id,
        error: {
          code: toolErrorCodes.UNKNOWN_TOOL,
          message: `unknown tool: ${params.name}`,
        },
      };
    }
    try {
      const out = await reg.handler(params.arguments);
      const result: ToolCallResult = { result: out };
      return { jsonrpc: JSONRPC_VERSION, id: req.id, result };
    } catch (err) {
      if (err instanceof ToolHandlerError) {
        return {
          jsonrpc: JSONRPC_VERSION,
          id: req.id,
          error: {
            code: toolErrorCodes.HANDLER_ERROR,
            message: err.message,
            data: err.data,
          },
        };
      }
      return {
        jsonrpc: JSONRPC_VERSION,
        id: req.id,
        error: { code: -32603, message: (err as Error).message ?? String(err) },
      };
    }
  }

  /**
   * Register a tool the host can invoke via orca/tools.call. Bare name
   * (no `<plugin_id>.` prefix — the host applies the namespace).
   * Re-registering the same name replaces the previous handler. Call
   * this for each tool, then call {@link declareTools} once.
   */
  registerTool(
    name: string,
    description: string,
    inputSchema: unknown,
    sensitivity: Sensitivity,
    handler: ToolHandler,
  ): void {
    const declaration: ToolDeclaration = {
      name,
      description,
      input_schema: inputSchema,
      sensitivity,
    };
    this.tools.set(name, { declaration, handler });
  }

  /**
   * Send the registered tool set via orca/tools.declare. Returns the
   * namespaced ids the host accepted. Idempotent — calling again
   * replaces the host-side set.
   */
  async declareTools(): Promise<ToolsDeclareResult> {
    const tools = Array.from(this.tools.values()).map(t => t.declaration);
    const params: ToolsDeclareParams = { tools };
    const resp = await this.call(TOOLS_DECLARE_METHOD, params);
    if (resp.error) {
      throw new Error(`${TOOLS_DECLARE_METHOD} rejected: ${resp.error.message}`);
    }
    return resp.result as ToolsDeclareResult;
  }

  /** Subscribe to every server-pushed notification on this connection. */
  notifications(handler: (n: Notification) => void): () => void {
    this.notifSubs.add(handler);
    return () => this.notifSubs.delete(handler);
  }

  /** Send a request and resolve with the matching response. */
  async call(method: string, params?: unknown): Promise<Response> {
    if (this.closed) throw new Error('transport closed');
    const id = this.nextID++;
    const req: Request = { jsonrpc: JSONRPC_VERSION, id, method, params };
    const promise = new Promise<Response>(resolve => this.pending.set(id, resolve));
    const body = Buffer.from(JSON.stringify(req), 'utf8');
    try {
      await writeFrame(this.writer, body);
    } catch (err) {
      this.pending.delete(id);
      throw err;
    }
    return promise;
  }

  async hello(
    pluginID: string,
    flavor: Flavor = 'headless',
    methodsRequired: string[] = [],
    methodsOptional: string[] = [],
  ): Promise<HelloResult> {
    const params: HelloParams = {
      sdk_version: SDK_VERSION,
      plugin_id: pluginID,
      flavor,
      core_min_required: '0.1.0',
      methods_required: methodsRequired,
      methods_optional: methodsOptional,
    };
    const resp = await this.call('orca/hello', params);
    if (resp.error) throw new Error(`orca/hello rejected: ${resp.error.message}`);
    const result = resp.result as HelloResult;
    if (!result.ok) {
      throw new Error(
        `orca/hello: server returned ok=false (status=${result.status}; ${result.reason ?? 'no reason'})`,
      );
    }
    return result;
  }

  async declareTypes(types: TypeDeclaration[]): Promise<TypesDeclareResult> {
    const resp = await this.call('orca/types.declare', { types });
    if (resp.error) throw new Error(`orca/types.declare rejected: ${resp.error.message}`);
    return resp.result as TypesDeclareResult;
  }

  async publishContext(contextID: string, value: TypedValue): Promise<void> {
    const resp = await this.call('orca/context.publish', { context_id: contextID, value });
    if (resp.error) throw new Error(`orca/context.publish rejected: ${resp.error.message}`);
  }

  async subscribeContext(
    contextID: string,
    typeFilter: string[] = [],
  ): Promise<{ subscriptionID: string; events: AsyncIterable<ContextEvent> }> {
    const resp = await this.call('orca/context.subscribe', {
      context_id: contextID,
      type_filter: typeFilter,
    });
    if (resp.error) throw new Error(`orca/context.subscribe rejected: ${resp.error.message}`);
    const subscriptionID = (resp.result as { subscription_id: string }).subscription_id;

    const queue: ContextEvent[] = [];
    const waiters: Array<(ev: IteratorResult<ContextEvent>) => void> = [];
    let done = false;

    const unsub = this.notifications(n => {
      if (n.method !== CONTEXT_EVENT_METHOD) return;
      const ev = n.params as ContextEvent;
      if (ev.subscription_id !== subscriptionID) return;
      if (waiters.length > 0) {
        waiters.shift()!({ value: ev, done: false });
      } else {
        queue.push(ev);
      }
    });

    const events: AsyncIterable<ContextEvent> = {
      [Symbol.asyncIterator](): AsyncIterator<ContextEvent> {
        return {
          next() {
            if (queue.length > 0) {
              return Promise.resolve({ value: queue.shift()!, done: false });
            }
            if (done) {
              return Promise.resolve({ value: undefined, done: true });
            }
            return new Promise(resolve => waiters.push(resolve));
          },
          return() {
            done = true;
            unsub();
            for (const w of waiters) w({ value: undefined, done: true });
            waiters.length = 0;
            return Promise.resolve({ value: undefined, done: true });
          },
        };
      },
    };
    return { subscriptionID, events };
  }

  async unsubscribeContext(subscriptionID: string): Promise<void> {
    const resp = await this.call('orca/context.unsubscribe', { subscription_id: subscriptionID });
    if (resp.error) throw new Error(`orca/context.unsubscribe rejected: ${resp.error.message}`);
  }
}

export type { ErrorObject };
